File: src/main/scala/fortress/transformers/SkolemizeTransformer.java
package fortress.transformers;

import fortress.tfol.*;
import java.util.Set;
import java.util.HashSet;
import fortress.data.NameGenerator;
import fortress.data.SubIntNameGenerator;
import fortress.tfol.operations.Skolemizer;

/**
* Takes a theory, whose formulas are all in negation normal form, and produces
* an equisatisfiable theory whose formulas are still in negation normal form and
* contain no existential quantifiers.
* Prenex normal form is preserved.
* Scopes are preserved.
*/
public class SkolemizeTransformer implements TheoryTransformer {
    @Override
    public Theory apply(Theory theory) {
        Set<String> forbiddenNames = new HashSet<>();
        
        for(Type type : theory.getTypes()) {
            forbiddenNames.add(type.getName());
        }
        
        for(FuncDecl fdecl : theory.getFunctionDeclarations()) {
            forbiddenNames.add(fdecl.getName());
        }
        
        for(AnnotatedVar c : theory.getConstants()) {
            forbiddenNames.add(c.getName());
        }
        
        // TODO: do we need this restriction if Substituter already restricts these inside one term?
        for(Term axiom : theory.getAxioms()) {
            forbiddenNames.addAll(axiom.allSymbolsJava());
        }
        
        NameGenerator nameGenerator = new SubIntNameGenerator(forbiddenNames, 0);
        
        Theory result = theory.withoutAxioms();
        for(Term axiom : theory.getAxioms()) {
            Skolemizer skolemizer = new Skolemizer(axiom, result.getSignature(), nameGenerator);
            Term newAxiom = skolemizer.convert();
            result = result.withFunctionDeclarations(skolemizer.getSkolemFunctions());
            result = result.withConstants(skolemizer.getSkolemConstants());
            result = result.withAxiom(newAxiom);
        }
        
        return result;
    }
    
    @Override
    public String getName() {
        return "Skolemize Transformer";
    }
}

File: src/main/scala/fortress/transformers/SkolemizeTransformer.scala
package fortress.transformers

import fortress.msfol._
import fortress.data.NameGenerator
import fortress.data.IntSuffixNameGenerator
import fortress.operations.Skolemizer
import fortress.operations.TermOps._
import fortress.operations.TheoryOps._
import fortress.modelfind.ProblemState

/** Given a theory, with formulas all in negation normal form, eliminates existential
* quantifiers through skolemization. The resulting theory is equisatisfiable with
* the original.
* Precondition: Input theory is in negation normal form.*/
class SkolemizeTransformer extends ProblemStateTransformer {
    
    override def apply(problemState: ProblemState): ProblemState = problemState match {
        case ProblemState(theory, scopes, skc, skf, unapplyInterp) => {
            val forbiddenNames = scala.collection.mutable.Set[String]()
            
            for(sort <- theory.sorts) {
                forbiddenNames += sort.name
            }
            
            for(fdecl <- theory.functionDeclarations) {
                forbiddenNames += fdecl.name
            }
            
            for(constant <- theory.constants) {
                forbiddenNames += constant.name
            }
            
            // TODO: do we need this restriction if Substituter already restricts these inside one term?
            for(axiom <- theory.axioms) {
                forbiddenNames ++= axiom.allSymbols
            }
            
            val nameGenerator = new IntSuffixNameGenerator(forbiddenNames.toSet, 0)
            
            var resultTheory = theory.withoutAxioms
            val newSkolemConstants = scala.collection.mutable.Set.empty[AnnotatedVar]
            val newSkolemFunctions = scala.collection.mutable.Set.empty[FuncDecl]
            for(axiom <- theory.axioms) {
                val skolemizer = new Skolemizer(axiom, resultTheory.signature, nameGenerator)
                val newAxiom = skolemizer.convert()
                newSkolemConstants ++= skolemizer.getSkolemConstants
                newSkolemFunctions ++= skolemizer.getSkolemFunctions
                resultTheory = resultTheory.withFunctionDeclarations(skolemizer.getSkolemFunctions.toList)
                resultTheory = resultTheory.withConstants(skolemizer.getSkolemConstants.toList)
                resultTheory = resultTheory.withAxiom(newAxiom)
            }
            
            ProblemState(resultTheory, scopes, skc ++ newSkolemConstants.toSet, skf ++ newSkolemFunctions.toSet, unapplyInterp)
        }
    }
    
    override def name: String = "Skolemize Transformer"
    
}
